src/cursor/cur_version.c

```diff
@@ -0,0 +1,387 @@
+                  WT_VERSION_UPDATE_CHAIN);
+
+                /*
+                 * Set the version cursor's value to be the update value. If the update is a modify,
+                 * reconstruct the value.
+                 */
+                if (upd->type != WT_UPDATE_MODIFY)
+                    __wt_upd_value_assign(cbt->upd_value, upd);
+                else
+                    WT_RET(
+                      __wt_modify_reconstruct_from_upd_list(session, cbt, upd, cbt->upd_value));
+
+                __wt_cursor_set_value(cursor, cbt->upd_value->buf);
+
+                upd = upd->next;
+                if (upd == NULL)
+                    F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+                goto done;
+            }
+        }
+    }
+
+    if (!F_ISSET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED)) {
+        WT_RET(__wt_value_return_buf(cbt, cbt->ref, &cbt->upd_value->buf, &tw));
+
+        if (!WT_TIME_WINDOW_HAS_STOP(&tw)) {
+            ondisk_durable_stop_ts = version_cursor->upd_durable_stop_ts;
+            ondisk_stop_ts = version_cursor->upd_stop_ts;
+            ondisk_stop_txn = version_cursor->upd_txnid;
+        } else {
+            ondisk_durable_stop_ts = tw.durable_stop_ts;
+            ondisk_stop_ts = tw.stop_ts;
+            ondisk_stop_txn = tw.stop_txn;
+        }
+
+        __wt_cursor_set_key(cursor, tw.start_txn, tw.durable_start_ts, tw.start_ts, ondisk_stop_txn,
+          ondisk_durable_stop_ts, ondisk_stop_ts, 0, 0, 0, WT_VERSION_DISK_IMAGE);
+        __wt_cursor_set_value(cursor, cbt->upd_value->buf);
+
+        version_cursor->upd_txnid = tw.start_txn;
+        version_cursor->upd_durable_stop_ts = tw.durable_start_ts;
+        version_cursor->upd_stop_ts = tw.start_ts;
+
+        F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+        goto done;
+    }
+
+    if (!F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, cursor->key, WT_TS_MAX, UINT64_MAX);
+        ret = __wt_curhs_search_near_before(session, hs_cursor);
+        ret = hs_cursor->prev(hs_cursor);
+
+        /*
+         * If there are no history store records for the given key or if we have iterated through
+         * all the records already, we have exhausted the history store.
+         */
+        if (ret == 0) {
+            WT_TIME_WINDOW_INIT(&tw);
+            hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &tw.start_ts, &hs_counter);
+            hs_cursor->get_value(
+              hs_cursor, &tw.stop_ts, &tw.durable_start_ts, &hs_upd_type, hs_value);
+            __wt_cursor_set_key(cursor, tw.start_txn, tw.durable_start_ts, tw.start_ts,
+              version_cursor->upd_txnid, version_cursor->upd_durable_stop_ts,
+              version_cursor->upd_stop_ts, hs_upd_type, 0, 0, WT_VERSION_HISTORY_STORE);
+            __wt_cursor_set_value(cursor, hs_value->data);
```

Comment:
This is wrong. You need to resolve reverse modify if needed. hs may store reverse modifies not full value. Construct the full value if needed.